File: 2025-07-July-GFG-POTD/July 17 - Power of k in factorial of n/Solution.py
class Solution:
    def maxKPower(self, n: int, k: int) -> int:
        # Factorize k into its prime factors
        prime_factors = {}
        num = k
        i = 2
        while i * i <= num:
            while k % i == 0:
                prime_factors[i] = prime_factors.get(i, 0) + 1
                k //= i
            i += 1
        if k > 1:
            prime_factors[k] = prime_factors.get(k, 0) + 1

        # Apply Legendre's formula
        result = float('inf')
        for p, a in prime_factors.items():
            count = 0
            temp = n
            while temp > 0:
                temp //= p
                count += temp
            result = min(result, count // a)

        return result

File: 2025-07-July-GFG-POTD/July 17 - Power of k in factorial of n/Solution.java
class Solution {
    public int maxKPower(int n, int k) {
        // Factorize k into its prime factors
        Map<Integer, Integer> primeFactors = new HashMap<>();
        int num = k;
        for (int i = 2; i * i <= num; i++) {
            while (k % i == 0) {
                primeFactors.put(i, primeFactors.getOrDefault(i, 0) + 1);
                k /= i;
            }
        }
        if (k > 1) {
            primeFactors.put(k, primeFactors.getOrDefault(k, 0) + 1);
        }

        // For each prime factor, use Legendre's formula
        int result = Integer.MAX_VALUE;
        for (Map.Entry<Integer, Integer> entry : primeFactors.entrySet()) {
            int p = entry.getKey();
            int a = entry.getValue();
            int count = 0;
            int temp = n;
            while (temp > 0) {
                temp /= p;
                count += temp;
            }
            result = Math.min(result, count / a);
        }

        return result;
    }

}
